great_expectations/profile/user_configurable_profiler.py

```diff
@@ -0,0 +1,1122 @@
+                    values contain fewer than 50% null values, then the profiler will add
+                    `expect_column_values_to_not_be_null`; if greater than 50% it will add
+                    `expect_column_values_to_be_null`. If not_null_only is set to True, the profiler will add a
+                    not_null expectation irrespective of the percent nullity (and therefore will not add an
+                    `expect_column_values_to_be_null`
+                "primary_or_compound_key": A list containing one or more columns which are a dataset's primary or
+                    compound key. This will create an `expect_column_values_to_be_unique` or
+                    `expect_compound_columns_to_be_unique` expectation. This will occur even if one or more of the
+                    primary_or_compound_key columns are specified in ignored_columns
+                "semantic_types": A dictionary where the keys are available semantic_types (see ProfilerSemanticTypes)
+                    and the values are lists of columns for which you would like to create semantic_type specific
+                    expectations e.g.:
+                    "semantic_types": { "value_set": ["state","country"], "numeric":["age", "amount_due"]}
+                "table_expectations_only": Boolean, default False. If True, this will only create the two table level expectations
+                    available to this profiler (`expect_table_columns_to_match_ordered_list` and
+                    `expect_table_row_count_to_be_between`). If a primary_or_compound key is specified, it will create
+                    a uniqueness expectation for that column as well
+                "value_set_threshold": Takes a string from the following ordered list - "none", "one", "two",
+                    "very_few", "few", "many", "very_many", "unique". When the profiler runs withou a semantic_types
+                    dict, each column is profiled for cardinality. This threshold determines the greatest cardinality
+                    for which to add `expect_column_values_to_be_in_set`. For example, if value_set_threshold is set to
+                    "unique", it will add a value_set expectation for every included column. If set to "few", it will
+                    add a value_set expectation for columns whose cardinality is one of "one", "two", "very_few" or
+                    "few". The default value is "many". For the purposes of comparing whether two tables are identical,
+                    it might make the most sense to set this to "unique"
+        """
+        self.dataset = dataset
+        self.column_info = {}
+
+        self.semantic_type_dict = semantic_types_dict
+        assert isinstance(self.semantic_type_dict, (dict, type(None)))
+
+        self.ignored_columns = ignored_columns or []
+        assert isinstance(self.ignored_columns, list)
```

Comment:
I love these checks to make sure parameters are the correct types!
You might want to take a look at the pattern used for other configs, e.g. creating a marshmallow schema to validate data structures e.g. DataContextConfigSchema: https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/data_context/types/base.py#L722
What you have here works, and this might be out of scope / overkill for the current effort. I just wanted to bring it up since I was recently working in base.py and noticed how we handle other configs. It would require creating a config class like `ProfilerConfig` and a schema like `ProfilerConfigSchema`.

Reply:
Great idea! I don't love the current pattern of making assertions for each parameter, and that sounds like an improvement. I think something like that is presently out of scope, but that's a great suggestion for future improvements.